File: server/src/Models/DMToolkitModels/Definitions/SubclassDefinition.cs
using DMToolkit.Models.JoinTables;

namespace DMToolkit.Models.Definitions;

public class SubclassDefinition : IDefinition
{
    public string Id { get; set; } = Guid.NewGuid().ToString();
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public ICollection<SubclassDefinitionFeatDefinition> SubclassDefinitionFeatDefinitions { get; set; } = new List<SubclassDefinitionFeatDefinition>();
    public string CharacterClassDefinitionId { get; set; } = string.Empty;
    public CharacterClassDefinition? CharacterClassDefinition { get; set; } = null;
}

File: flutter-client/dm_toolkit/lib/models/dm_toolkit/definitions/subclass_definition.dart
import 'package:dm_toolkit/models/dm_toolkit/definitions/feat_definition.dart';
import 'package:dm_toolkit/models/dm_toolkit/join_tables/feat_group_level.dart';

class SubclassDefinition {
  final String id;
  final String name;
  final String description;
  final List<FeatGroupLevel> featTables;

  SubclassDefinition({
    required this.id,
    required this.name,
    required this.description,
    required this.featTables
  });

  factory SubclassDefinition.fromJson(Map<String, dynamic> json, List<FeatDefinition> featDefinitions) {
    try {
      var featTableListJson = json['featTables'] as List;
      var featTables = featTableListJson
        .map((featTableJson) => FeatGroupLevel.fromJson(featTableJson, featDefinitions))
        .toList();
      
      return SubclassDefinition(
        id: json['id'] as String, 
        name: json['name'] as String, 
        description: json['description'] as String, 
        featTables: featTables
      );
    } catch (e) {
      throw FormatException('SubclassDefinition model is invalid.', e);
    } 
  }
}